### DISCUSSION TITLE
no scalar implementation for the named  'Date' scalar type

### DISCUSSION THREAD

**Discussion by mmriyadh** [graphql-java-kickstart/graphql-spring-boot#573] [Q&A] [answered]
I have added the @Bean definition as suggested [here](https://github.com/graphql-java-kickstart/graphql-spring-boot/issues/32#issuecomment-329180868). However, I'm getting this error below during spring boot server startup. I have generated the code from a schema (which contains `scalar Date`) using Netflix code generation [tool](https://github.com/Netflix/dgs-codegen).

`There is no scalar implementation for the named  'Date' scalar type`

**Detail error stack**
`2021-04-09 00:09:08.766 ERROR 62316 --- [           main] o.s.b.web.embedded.tomcat.TomcatStarter  : Error starting Tomcat context. Exception: org.springframework.beans.factory.UnsatisfiedDependencyException. Message: Error creating bean with name 'graphQLServletRegistrationBean' defined in class path resource [graphql/kickstart/spring/web/boot/GraphQLWebAutoConfiguration.class]: Unsatisfied dependency expressed through method 'graphQLServletRegistrationBean' parameter 0; nested exception is org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'graphQLHttpServlet' defined in class path resource [graphql/kickstart/spring/web/boot/GraphQLWebAutoConfiguration.class]: Unsatisfied dependency expressed through method 'graphQLHttpServlet' parameter 0; nested exception is org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'graphQLServletConfiguration' defined in class path resource [graphql/kickstart/spring/web/boot/GraphQLWebAutoConfiguration.class]: Unsatisfied dependency expressed through method 'graphQLServletConfiguration' parameter 0; nested exception is org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'invocationInputFactory' defined in class path resource [graphql/kickstart/spring/web/boot/GraphQLWebAutoConfiguration.class]: Unsatisfied dependency expressed through method 'invocationInputFactory' parameter 0; nested exception is org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'graphQLSchemaProvider' defined in class path resource [graphql/kickstart/spring/web/boot/GraphQLWebAutoConfiguration.class]: Unsatisfied dependency expressed through method 'graphQLSchemaProvider' parameter 0; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'schema' defined in class path resource [com/netflix/graphql/dgs/autoconfig/DgsAutoConfiguration.class]: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [graphql.schema.GraphQLSchema]: Factory method 'schema' threw exception; nested exception is SchemaProblem{errors=[There is no scalar implementation for the named  'Date' scalar type]}`

The dependency tree related to spring/graphql looks like this.

```
[INFO] +- org.springframework.boot:spring-boot-starter-web:jar:2.4.2:compile
[INFO] |  +- org.springframework.boot:spring-boot-starter:jar:2.4.2:compile
[INFO] |  |  +- org.springframework.boot:spring-boot:jar:2.4.2:compile
[INFO] |  |  +- org.springframework.boot:spring-boot-autoconfigure:jar:2.4.2:compile
[INFO] |  |  +- org.springframework.boot:spring-boot-starter-logging:jar:2.4.2:compile
[INFO] |  |  |  +- org.apache.logging.log4j:log4j-to-slf4j:jar:2.13.3:compile
[INFO] |  |  |  |  \- org.apache.logging.log4j:log4j-api:jar:2.13.3:compile
[INFO] |  |  |  \- org.slf4j:jul-to-slf4j:jar:1.7.30:compile
[INFO] |  |  +- jakarta.annotation:jakarta.annotation-api:jar:1.3.5:compile
[INFO] |  |  \- org.yaml:snakeyaml:jar:1.27:compile
[INFO] |  +- org.springframework.boot:spring-boot-starter-json:jar:2.4.2:compile
[INFO] |  |  +- com.fasterxml.jackson.core:jackson-databind:jar:2.11.4:compile
[INFO] |  |  |  \- com.fasterxml.jackson.core:jackson-annotations:jar:2.11.4:compile
[INFO] |  |  +- com.fasterxml.jackson.datatype:jackson-datatype-jdk8:jar:2.11.4:compile
[INFO] |  |  +- com.fasterxml.jackson.datatype:jackson-datatype-jsr310:jar:2.11.4:compile
[INFO] |  |  \- com.fasterxml.jackson.module:jackson-module-parameter-names:jar:2.11.4:compile
[INFO] |  +- org.springframework.boot:spring-boot-starter-tomcat:jar:2.4.2:compile
[INFO] |  |  +- org.apache.tomcat.embed:tomcat-embed-core:jar:9.0.41:compile
[INFO] |  |  +- org.glassfish:jakarta.el:jar:3.0.3:compile
[INFO] |  |  \- org.apache.tomcat.embed:tomcat-embed-websocket:jar:9.0.41:compile
[INFO] |  +- org.springframework:spring-web:jar:5.3.3:compile
[INFO] |  |  \- org.springframework:spring-beans:jar:5.3.3:compile
[INFO] |  \- org.springframework:spring-webmvc:jar:5.3.3:compile
[INFO] |     +- org.springframework:spring-aop:jar:5.3.3:compile
[INFO] |     +- org.springframework:spring-context:jar:5.3.3:compile
[INFO] |     \- org.springframework:spring-expression:jar:5.3.3:compile
[INFO] +- org.springframework.boot:spring-boot-starter-actuator:jar:2.4.2:compile
[INFO] |  +- org.springframework.boot:spring-boot-actuator-autoconfigure:jar:2.4.2:compile
[INFO] |  |  \- org.springframework.boot:spring-boot-actuator:jar:2.4.2:compile
[INFO] |  \- io.micrometer:micrometer-core:jar:1.6.3:compile
[INFO] |     +- org.hdrhistogram:HdrHistogram:jar:2.1.12:compile
[INFO] |     \- org.latencyutils:LatencyUtils:jar:2.0.3:runtime
[INFO] +- org.springframework.boot:spring-boot-configuration-processor:jar:2.4.2:compile (optional) 
[INFO] +- com.graphql-java-kickstart:graphql-spring-boot-starter:jar:8.0.0:compile
[INFO] |  \- com.graphql-java-kickstart:graphql-spring-boot-autoconfigure:jar:8.0.0:compile
[INFO] |     +- com.graphql-java-kickstart:graphql-kickstart-spring-boot-starter-tools:jar:8.0.0:compile
[INFO] |     |  \- com.graphql-java-kickstart:graphql-kickstart-spring-boot-autoconfigure-tools:jar:8.0.0:compile
[INFO] |     |     \- com.graphql-java-kickstart:graphql-java-tools:jar:6.2.0:compile
[INFO] |     |        +- org.jetbrains.kotlinx:kotlinx-coroutines-jdk8:jar:1.4.2:compile
[INFO] |     |        +- org.jetbrains.kotlinx:kotlinx-coroutines-core:jar:1.4.2:compile
[INFO] |     |        +- org.jetbrains.kotlinx:kotlinx-coroutines-reactive:jar:1.4.2:compile
[INFO] |     |        \- com.fasterxml:classmate:jar:1.5.1:compile
[INFO] |     +- com.graphql-java-kickstart:graphql-kickstart-spring-support:jar:8.0.0:compile
[INFO] |     +- com.graphql-java-kickstart:graphql-java-kickstart:jar:10.0.0:compile
[INFO] |     +- com.graphql-java-kickstart:graphql-java-servlet:jar:10.0.0:compile
[INFO] |     |  +- javax.servlet:javax.servlet-api:jar:4.0.1:compile
[INFO] |     |  \- javax.websocket:javax.websocket-api:jar:1.1:compile
[INFO] |     \- com.graphql-java:graphql-java:jar:15.0:compile
[INFO] |        +- org.antlr:antlr4-runtime:jar:4.7.2:compile
[INFO] |        +- com.graphql-java:java-dataloader:jar:2.2.3:compile
[INFO] |        \- org.reactivestreams:reactive-streams:jar:1.0.3:compile
[INFO] +- com.netflix.graphql.dgs:graphql-dgs-spring-boot-starter:jar:3.4.0:compile
[INFO] |  +- com.netflix.graphql.dgs:graphql-dgs-spring-boot-oss-autoconfigure:jar:3.4.0:compile
[INFO] |  |  +- com.netflix.graphql.dgs:graphql-dgs:jar:3.4.0:compile
[INFO] |  |  |  +- com.netflix.graphql.dgs:graphql-dgs-mocking:jar:3.4.0:compile
[INFO] |  |  |  |  \- com.github.javafaker:javafaker:jar:1.0.2:runtime
[INFO] |  |  |  |     +- org.yaml:snakeyaml:jar:android:1.23:runtime
[INFO] |  |  |  |     \- com.github.mifmif:generex:jar:1.0.2:runtime
[INFO] |  |  |  |        \- dk.brics.automaton:automaton:jar:1.11-8:runtime
[INFO] |  |  |  +- com.apollographql.federation:federation-graphql-java-support:jar:0.5.0:runtime
[INFO] |  |  |  |  \- com.google.protobuf:protobuf-java:jar:3.9.0:runtime
[INFO] |  |  |  \- org.springframework.security:spring-security-core:jar:5.4.2:runtime
[INFO] |  |  +- com.netflix.graphql.dgs:graphql-dgs-spring-webmvc:jar:3.4.0:compile
[INFO] |  |  |  \- org.jetbrains.kotlin:kotlin-reflect:jar:1.4.21:compile
[INFO] |  |  \- org.jetbrains.kotlin:kotlin-stdlib-jdk8:jar:1.4.21:compile
[INFO] |  |     +- org.jetbrains.kotlin:kotlin-stdlib:jar:1.4.21:compile
[INFO] |  |     |  +- org.jetbrains.kotlin:kotlin-stdlib-common:jar:1.4.21:compile
[INFO] |  |     |  \- org.jetbrains:annotations:jar:13.0:compile
[INFO] |  |     \- org.jetbrains.kotlin:kotlin-stdlib-jdk7:jar:1.4.21:compile
[INFO] |  +- com.netflix.graphql.dgs:graphql-dgs-spring-webmvc-autoconfigure:jar:3.4.0:compile
[INFO] |  +- com.netflix.graphql.dgs:graphql-dgs-client:jar:3.4.0:compile
[INFO] |  |  +- io.projectreactor:reactor-core:jar:3.4.2:compile
[INFO] |  |  \- com.fasterxml.jackson.module:jackson-module-kotlin:jar:2.11.4:compile
[INFO] |  +- com.netflix.graphql.dgs:graphql-error-types:jar:3.4.0:compile
[INFO] |  +- com.netflix.graphql.dgs:graphql-dgs-graphiql-autoconfigure:jar:3.4.0:runtime
[INFO] |  \- org.springframework.boot:spring-boot-starter-websocket:jar:2.4.2:compile
[INFO] |     +- org.springframework:spring-messaging:jar:5.3.3:compile
[INFO] |     \- org.springframework:spring-websocket:jar:5.3.3:compile
```

This is the java file containing the main method.

```
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
import graphql.kickstart.tools.ObjectMapperConfigurer;
import graphql.kickstart.tools.ObjectMapperConfigurerContext;
import graphql.kickstart.tools.SchemaParserOptions;
import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.context.annotation.Bean;

@SpringBootApplication
public class ExampleApplication {

  public static void main(String[] args) {
    SpringApplication.run(ExampleApplication.class, args);
  }

  // https://github.com/graphql-java-kickstart/graphql-spring-boot/issues/32
  @Bean
  ObjectMapperConfigurer objectMapperConfigurer() {
    return (ObjectMapper mapper, ObjectMapperConfigurerContext context) -> mapper
        .registerModule(new JavaTimeModule());
  }
```

**Comment by oliemansm** [comment, marked as answer]
You seem to be registering the "wrong" bean. The stack trace is clearly complaining about it not being able to find the scalar type. That's because you need to expose that scalar class as a bean. See also [this answer](https://github.com/graphql-java-kickstart/graphql-spring-boot/discussions/460#discussioncomment-225852).

I don't know what that codegen tool generated for you, since I don't have any experience with it. My guess is it generated something like the `DateCoercing` as used below. Not sure how it would do that tbh, since those classes require certain logic that I can't imagine a tool generating for you. 

```java
  @Bean
  public GraphQLScalarType dateScalar() {
    return GraphQLScalarType.newScalar()
        .name("Date")
        .description("Date type")
        .coercing(new DateCoercing())
        .build();
  }
```

Does that Netflix tool advertise that the generated classes are compatible with the `graphql-spring-boot-starter` icw `graphql-java-tools`? And if so, do they provide documentation how that should work? If not, I hope the above will fix it for you, but if not I'd advise you to first dive into this library without that generated code if you don't have any experience with this library yet. Because in that case it seems you would have to make some tweaks to make that generated code compatible with this library.

**Reply by mmriyadh** [reply]
Thanks for your response. After digging through some sample code from Netflix, it appears that I need to add this class for autowiring with spring.

```
import com.netflix.graphql.dgs.DgsComponent;
import com.netflix.graphql.dgs.DgsRuntimeWiring;
import graphql.scalars.ExtendedScalars;
import graphql.schema.idl.RuntimeWiring;

/**
 * graphql-java provides optional scalars in the graphql-java-extended-scalars library.
 * We can wire a scalar from this library by adding the scalar to the RuntimeWiring.
 */
@DgsComponent
public class DateScalar {

  @DgsRuntimeWiring
  public RuntimeWiring.Builder addScalar(RuntimeWiring.Builder builder) {
    return builder.scalar(ExtendedScalars.Date);
  }
}

```

**Reply by oliemansm** [reply]
Right, so it has nothing to do with this library at all. That tool and library uses graphql-java directly. You probably don't need any of the packages from this project in the first place.
